### ISSUE TITLE
Improvements/clarification around msbuild's parallel flag

### ISSUE THREAD

**Issue by malwoden** [conan-io/conan#2699]
- [x] I've read the [CONTRIBUTING guide](https://raw.githubusercontent.com/conan-io/conan/develop/.github/CONTRIBUTING.md).
- [x] I've specified the Conan version, operating system version and any tool that can be relevant.
- [x] I've explained the steps to reproduce the error or the motivation/use case of the question/suggestion.

Conan 1.2, Windows 10 x64, msvc toolchain (using msvc15, but I believe this is all relevant down to v10).

## Background/Overview

The MSBuild build helper allows a 'parallel' flag to be set when executing a build:
`msbuild.build("My.sln", parallel=True)`

This results in the /m ([msdn doc](https://msdn.microsoft.com/en-us/library/bb651793.aspx)) flag being passed to msbuild:
`msbuild My.sln /p:Configuration=Release /p:Platform="x86" /m:4`

This allows msbuild to build multiple projects in parallel, however the actual compilation of a single project will be done serially. This means that if you have a single target, or many target that are all linearly dependent (a depends on b depends on c), then you won't see a speed up when using this flag.

To allow parallel compilation of source files within a single project we have to pass the MP parameter to CL.exe ([msdn doc](https://msdn.microsoft.com/en-us/library/bb385193.aspx)). This can be done currently with conan by setting the CL environment variable before invoking msbuild:

```
with tools.environment_append({"CL": "/MP%s" % str(cpu_count())}):
  msbuild.build("My.sln")
```

This is the similar to calling `make -jN` with gnumake on a single target. This flag can be passed from msbuild to CL by using the `CL_MPCount` flag ([old msdn blog on this flag](https://blogs.msdn.microsoft.com/visualstudio/2010/03/07/tuning-c-build-parallelism-in-vs2010/))

## How conan can improve:
I was slightly surprised at first when I saw that my 'parallel' flag was set to True on the call to `msbuild::build` but I was still only using a single CPU core for the compiling.

The problem is that there are no perfect general values for the /m and /MP flags that suit all projects, it depends on how many projects can be built concurrently in the solution file. 

With that in mind I have a few ideas for conan:
1. Update the documentation for the parallel flag to state that it is for parallel project building, not parallel source building.
2. Change parallel to set the CL /MP flag rather than the msbuild /m flag. This would be ideal for solutions with have a small number of projects with many source files, less ideal for solutions with lots of independent projects.
3. Allow the setting of parallel_project and parallel_source flags independently on the call to msbuild
4. Keep a single 'parallel' flag and find some other intelligent way of determining the ideal /m and /MP flags

Number 4 would be ideal though I am not sure how possible it is. Number 2 and 3 are more subjective, it would be great to get other people's opinions on this.

**Comment by memsharded** [comment]
Yes, I agree that the ``MSBuild`` helper should be able to improve over this, maybe set the MP flag, as the ``CMake`` helper is already doing (via CONAN_CXX_FLAGS). Other helpers use internally ``tools.cpu_count()``, which is also configurable via conan.conf and an environment variable, so conan users can decide how many cores are dedicated to building.

I don't believe in such "intelligent ways" to determine the ideal parallelism, I'd say that we should probably go for a reasonable default and then allow the helper to configure them at will for different projects (mabye as a function of ``tools.cpu_count()``?)

In summary, yes, the MSBuild helper is actually pretty new, and it can be improved.
